Automated Testing: Add GitHub Actions workflow for Storybook smoke testing

im3dabasia opened the pull request:
## What?
Closes https://github.com/WordPress/gutenberg/issues/41456

## Why?
Code changes can break Storybook stories without anyone noticing until much later. Since our Storybook contains realistic usage examples, broken stories may indicate issues for third-party consumers.

## How?
Added a GitHub Actions workflow that:

1. Runs on pull requests
2. Builds and serves Storybook
3. Uses @storybook/test-runner to check all stories render without errors
4. Properly manages the server process lifecycle

This catches breaking changes early before they're merged.

## Screenshots

This is the [Action Link](https://github.com/WordPress/gutenberg/actions/runs/14057695334/job/39360778872?pr=69679)
<img width="1440" alt="image" src="https://github.com/user-attachments/assets/a635e35e-44f2-48e8-b08f-3a1dd290b350" />

im3dabasia committed:
feat: Add test-runner and setup smoke tests for storybook

im3dabasia committed:
feat: Add build command

im3dabasia committed:
fix: Script command in package.json

im3dabasia committed:
restore: package-lock from trunk

im3dabasia committed:
restore: Old package.json

im3dabasia committed:
feat: Add storybook package

im3dabasia committed:
fix: Add node in types

im3dabasia committed:
remove: test-runner dep

im3dabasia committed:
fix: Update the smoke test run command

im3dabasia committed:
feat: Dev server to work in bg mode, sleep logic

im3dabasia committed:
fix: yaml file run command

im3dabasia committed:
chore: Add playwright install command

im3dabasia commented:
Hi @mirka,

When you have a moment I'd like to hear your thoughts on this Enhancement on our CI/CD Automated testing.
Looking forward to your feedback

mirka reviewed (commented):
Yes, I think the overall approach of using `test-storybook` for automated smoke tests is solid!

I don't have the bandwidth to go super deep on reviewing/testing the workflow itself, but a few thoughts I had:

- Any reason we're doing the `pid` thing rather than something more like the official example [here](https://storybook.js.org/docs/writing-tests/test-runner#run-against-non-deployed-storybooks)?
- And if we do switch to the `build`-based approach like the example above, maybe we can save time by running the smoke tests in this [existing workflow](https://github.com/WordPress/gutenberg/blob/trunk/.github/workflows/storybook-check.yml)?

im3dabasia committed:
fix: Add concurrent exition logic when tests are run

im3dabasia committed:
fix: Increase stories timeout and added sleep

im3dabasia committed:
fix: Add timeout between jobs

im3dabasia committed:
chore: Add the smoke tests to storybook-check workflow

im3dabasia committed:
chore: Remove the storybook-test yaml

im3dabasia commented:
Thank you @mirka for your feedback.

I have implemented the necessary changes in this commit: https://github.com/WordPress/gutenberg/pull/69679/commits/06a6c1c296b83034b47974077bec7c942b33b2d8.

This is the related CI test action from the above commit: [Storybook build and Smoke Tests](https://github.com/WordPress/gutenberg/actions/runs/14126989524/job/39578177139?pr=69679)
 
I've now integrated the concurrently package to properly handle process termination when tests finish. The smoke tests have been incorporated into the existing `storybook-check.yml` file, and I've renamed the workflow to "Storybook Build and Smoke Tests" to better reflect its expanded functionality.

Since you mentioned you have limited bandwidth, I'll tag @Mamaduka and @t-hamano to continue the review process from here.

Thank you for your guidance!

mirka commented on .github/workflows/storybook-check.yml:
```diff
@@ -25,3 +26,19 @@ jobs:
+            - name: Serve Storybook and run tests
+              run: |
+                  npx concurrently -k -s first -n "SB,TEST" -c "magenta,blue" \
+                  "npm run storybook:dev" \
```
Can we just serve (`http-server`) the built storybook like in the [example](https://storybook.js.org/docs/writing-tests/test-runner#run-against-non-deployed-storybooks)? That is what I hoped would save time, instead of rebuilding with `storybook:dev`.

desrosj commented on .github/workflows/storybook-check.yml:
```diff
@@ -13,6 +13,7 @@ jobs:
     check:
         runs-on: ubuntu-latest
         if: ${{ github.repository == 'WordPress/gutenberg' || github.event_name == 'pull_request' }}
+        timeout-minutes: 60
```
Curious why this is so high. Looking at the workflow runs so far, these usually complete under 20 minutes with a few outliers around 25 minutes. Would a more conservative value of `30` or `40` work here?

desrosj commented on .github/workflows/storybook-check.yml:
```diff
@@ -25,3 +26,19 @@ jobs:
+            - name: Install Playwright dependencies
+              run: npx playwright install
+
+            - name: Serve Storybook and run tests
```
Can "serve" and "run tests" parts be split into separate steps? It can make it more clear what is actually failing when issues are encountered.

im3dabasia replied:
@desrosj  I did that previously, but I was advised that we should follow the example in the Storybook documentation [here](https://storybook.js.org/docs/writing-tests/test-runner#run-against-non-deployed-storybooks).

im3dabasia committed:
fix: Use http-server and removed sleep time

im3dabasia committed:
fix: Update the static path in the yaml

im3dabasia committed:
chore: Add with deps flag in playwright install

mirka reviewed (approved):
Ok, so all the Playwright tests we added here only seem to be [adding](https://github.com/WordPress/gutenberg/actions/workflows/storybook-check.yml) a couple minutes to the previous version of the Storybook workflow, which seems reasonable!

Let's keep an eye out for anomalies, and tweak or remove if this turns out to be more trouble than it's worth 😂 I'll preemptively suggest a possible optimization: Historically, certain sections of the Storybook have been more prone to breakage (e.g. Block Editor Playground), so if testing every story ends up being too costly, we could maybe limit it to the problem sections.

Thanks @im3dabasia for your persistence here!

im3dabasia commented:
Hey @mirka ,

Could we merge these and see how it performs? We'd get a fair idea on if any improvements are needed after testing it in the Pipeline. 

Let me know, if further changes are needed . 

Regards.
